### PR TITLE
[Core][Enable gcs scheduler 7/n] Prefer actor owner's node

### PR DESCRIPTION
**Pull request by Chong-Li**
Signed-off-by: Chong-Li <lc300133@antgroup.com>

## Why are these changes needed?
This PR tries to finalize gcs actor scheduler, with the following changes:
1. Similar to the legacy scheduler, gcs now prefers the actor owner's node. It's usually required by RL cases for better colocation.
2. The normal task workers report zero CPU resources (instead of the allocated one) when they are blocked.
3. Similar to the legacy scheduler, gcs now schedules empty-resource actors randomly.
4. A new release test is added: multiple masters/drivers are creating (slave) actors concurrently. The case is able to expose the difference between centralized (gcs-based, less scheduling conflicts) and distributed schedulers.

The feature flag is temporarily turned on when going through the CI pipline. We still need another dedicated PR to turn it on by default.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by Chong-Li** [comment]
The RLLib release test results with raylet-based scheduler: https://buildkite.com/ray-project/release-tests-pr/builds/22539#_

With gcs-based-scheduler: https://buildkite.com/ray-project/release-tests-pr/builds/22674#_

I did not find notable performance regression. But the variations may need @rkooo567 to double check.

File: python/ray/_private/ray_constants.py
Comment (wumuzi520) [inline python/ray/_private/ray_constants.py:362]: Does this mean that gcs actor scheduling is enabled by default?

File: src/ray/common/ray_config_def.h
Comment (wumuzi520) [inline src/ray/common/ray_config_def.h:605]: Does this mean that gcs actor scheduling is enabled by default?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (wumuzi520) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:53]: This condition is a little weird here.

**Review by wumuzi520 — APPROVED** [review]
It looks good overall.
BTW, don't forget to change the switch `RAY_gcs_actor_scheduling_enabled` back.

File: src/ray/common/ray_config_def.h
Comment (Chong-Li) [reply src/ray/common/ray_config_def.h:605]: It is only turned on to go through CI pipeline. Will turn it off after code review.

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (Chong-Li) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:53]: This is just a quick way to make the empty-resource actors go through the legacy logic (randomly select a forward raylet).

**Comment by Chong-Li** [comment]
Is there any more suggestions to this PR? @iycheng @rkooo567

File: python/ray/_private/ray_constants.py
Comment (rkooo567) [inline python/ray/_private/ray_constants.py:362]: revert?

File: release/nightly_tests/many_nodes_tests/multi_master_test.py
Comment (rkooo567) [inline release/nightly_tests/many_nodes_tests/multi_master_test.py:9]: have you run this test? (do you need help from us running this? )

File: release/nightly_tests/many_nodes_tests/multi_master_test.py
Comment (rkooo567) [inline release/nightly_tests/many_nodes_tests/multi_master_test.py:71]: can you tell me a bit more about the success / failure criteria of this test?

File: release/release_tests.yaml
Comment (rkooo567) [inline release/release_tests.yaml:4006]: can we test this in smaller # of nodes?

File: src/ray/common/ray_config_def.h
Comment (rkooo567) [inline src/ray/common/ray_config_def.h:605]: I will request for changes until this change is reverted

**Review by rkooo567 — CHANGES_REQUESTED** [review]
Request change until we revert the gcs actor scheduling enabled by default

I have mainly two questions.
1. Is there any way to merge "prioritize_local_node "and "preferred_node_id"? I feel like they are the same thing. 
2. Some questions regarding the nightly tests (follow up from comments)

Also I am waiting for the review from @scv119

File: release/nightly_tests/many_nodes_tests/multi_master_test.py
Comment (Chong-Li) [reply release/nightly_tests/many_nodes_tests/multi_master_test.py:71]: This test is similar to the existing one: https://github.com/ray-project/ray/blob/master/release/nightly_tests/many_nodes_tests/actor_test.py. They both focus on the scheduling latency of a large number of actors. 

The only difference from the `actor_test.py` is that this test simulates a multi-master(driver) scenario. This means (distributed) master actors are creating slave actors concurrently. Please see this discussion (https://ray-distributed.slack.com/archives/C0334JQHG86/p1667468940596309) for details.

File: release/release_tests.yaml
Comment (Chong-Li) [reply release/release_tests.yaml:4006]: This config is actually the same as the one of `actor_test.py`. They should have the same scale to make an apple-to-apple comparison.

File: src/ray/common/ray_config_def.h
Comment (Chong-Li) [reply src/ray/common/ray_config_def.h:605]: Sure. I was meaning to revert it after code review. But could do it right now.

**Comment by Chong-Li** [comment]
> Request change until we revert the gcs actor scheduling enabled by default
> 
> I have mainly two questions.
> 
> 1. Is there any way to merge "prioritize_local_node "and "preferred_node_id"? I feel like they are the same thing.
> 2. Some questions regarding the nightly tests (follow up from comments)
> 
> Also I am waiting for the review from @scv119

These two questions have been resolved. Are there any more comments? @rkooo567 @scv119

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:97]:
```suggestion
  const auto &owner_node = gcs_node_manager_.GetAliveNode(actor->GetOwnerNodeID());
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler.cc:257]: nice change :)!

File: src/ray/common/task/task.h
Comment (rkooo567) [inline src/ray/common/task/task.h:62]: Don't we need to support optional? It looks like this is always passed

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_task_manager.cc:196]: How does it work if there's no preferred node id?

File: src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/policy/hybrid_scheduling_policy.cc:41]:
```suggestion
  auto preferred_it = nodes_.find(preferred_node_id);
```

File: src/ray/raylet/scheduling/cluster_resource_scheduler_test.cc
Comment (rkooo567) [inline src/ray/raylet/scheduling/cluster_resource_scheduler_test.cc:417]: do we have an unit test to verify if actor is scheduled on a owner node?

**Review by rkooo567 — COMMENTED** [review]
Looks pretty good! Last couple comments.!

File: src/ray/raylet/local_task_manager.cc
Comment (Chong-Li) [reply src/ray/raylet/local_task_manager.cc:1025]: `resource_request` can not be const because it is updated below.

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (Chong-Li) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:196]: Empty preferred node_id is handled specially (if empty, fallback to local node) inside `HybridSchedulingPolicy`.

File: src/ray/raylet/scheduling/cluster_resource_scheduler_test.cc
Comment (Chong-Li) [reply src/ray/raylet/scheduling/cluster_resource_scheduler_test.cc:456]: I will add one.

File: src/ray/raylet/scheduling/cluster_task_manager.cc
Comment (rkooo567) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:196]: Is it possible to just pass nullopt instead? It makes sense, but I think it is pretty confusing from code to understand

Comment (Chong-Li) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:196]: Right now, `preferred_node` is just a `std::string` in `ClusterResourceScheduler`, `RayTask`, `SchedulingOptions` and `HybridSchedulingPolicy`. Do you mean changing it to `std::optional<std::string>` globally?

Comment (Chong-Li) [reply src/ray/raylet/scheduling/cluster_task_manager.cc:196]: As suggested, more comments about the empty string have been added. @rkooo567

**Review by rkooo567 — APPROVED** [review]
There are test failures

**Comment by rkooo567** [comment]
I started a mac build. Please wait a little!

**Comment by Chong-Li** [comment]
> I started a mac build. Please wait a little!

All tests passed. @rkooo567

**Comment by rkooo567** [comment]
cc @iycheng

**Pull request merged**
